File: 0349-intersection-of-two-arrays/0349-intersection-of-two-arrays.go
func intersection(nums1 []int, nums2 []int) []int {
    res := []int{}
    mp := make(map[int]int)

    for _, num := range nums1 {
        mp[num] = mp[num] + 1
    }

    for _, num := range nums2 {
        if _, exists := mp[num]; exists {
            res = append(res, num)
            delete(mp, num)
        }
    }

    return res
}

File: 0349-intersection-of-two-arrays/0349-intersection-of-two-arrays.rs
use std::collections::HashMap;

impl Solution {
    pub fn intersection(nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<i32> {
        let mut res = Vec::new();
        let mut mp = HashMap::new();

        for &num in &nums1 {
            *mp.entry(num).or_insert(0) += 1;
        }

        for &num in &nums2 {
            if mp.contains_key(&num) {
                res.push(num);
                mp.remove(&num);
            }
        }

        res
    }
}